File: app/src/main/java/in/completecourse/questionbank/CompetitionUpdatesActivity.java
package in.completecourse.questionbank;

import android.app.ProgressDialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import java.lang.ref.WeakReference;
import java.util.ArrayList;

import cz.msebera.android.httpclient.HttpResponse;
import cz.msebera.android.httpclient.client.HttpClient;
import cz.msebera.android.httpclient.client.methods.HttpPost;
import cz.msebera.android.httpclient.entity.StringEntity;
import cz.msebera.android.httpclient.impl.client.DefaultHttpClient;
import cz.msebera.android.httpclient.params.HttpConnectionParams;
import cz.msebera.android.httpclient.util.EntityUtils;
import in.completecourse.questionbank.adapter.CompetitionUpdatesAdapter;
import in.completecourse.questionbank.app.AppConfig;
import in.completecourse.questionbank.model.UpdateItem;

public class CompetitionUpdatesActivity extends AppCompatActivity {
    private ProgressDialog pDialog;
    private ArrayList<UpdateItem> updatesList;
    private CompetitionUpdatesAdapter adapter;
    private RecyclerView recyclerView;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_dialog);

        recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        String classStringFinal = SubjectActivity.classString;


        JSONObject data = new JSONObject();
        try {
            data.putOpt("classid", classStringFinal);
            GetUpdates jsonTransmitter = new GetUpdates(CompetitionUpdatesActivity.this);
            jsonTransmitter.execute(data);
        } catch (JSONException e) {
            e.printStackTrace();
        }



    }


    private static class GetUpdates extends AsyncTask<JSONObject, JSONObject, JSONObject> {
        private final WeakReference<CompetitionUpdatesActivity> activityWeakReference;
        UpdateItem item;

        GetUpdates(CompetitionUpdatesActivity context){
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected void onPreExecute() {
            CompetitionUpdatesActivity activity = activityWeakReference.get();
            activity.pDialog = new ProgressDialog(activity);
            activity.pDialog.setMessage("Please wait...");
            activity.pDialog.setCancelable(false);
            activity.pDialog.show();
        }

        @Override
        protected JSONObject doInBackground(JSONObject... data) {
            final CompetitionUpdatesActivity activity = activityWeakReference.get();
            JSONObject json = data[0];
            HttpClient client = new DefaultHttpClient();
            HttpConnectionParams.setConnectionTimeout(client.getParams(), 100000);
            JSONObject jsonResponse;
            String LOGIN_URL = AppConfig.URL_COMPETITION_UPDATES;
            HttpPost post = new HttpPost(LOGIN_URL);
            try {
                StringEntity se = new StringEntity( json.toString());
                post.addHeader("content-type", "application/json");
                post.addHeader("accept", "application/json");
                post.setEntity(se);
                HttpResponse response;
                response = client.execute(post);
                String resFromServer = EntityUtils.toString(response.getEntity());
                jsonResponse = new JSONObject(resFromServer);
                activity.updatesList = new ArrayList<>();

                if (!jsonResponse.has("message")){
                    JSONObject jsonObject = new JSONObject(resFromServer);
                    JSONArray jsonArray = jsonObject.getJSONArray("data");
                    for (int i = 0; i<jsonArray.length(); i++) {
                        item =  new UpdateItem();
                        JSONObject chapterObject = jsonArray.getJSONObject(i);
                        item.setUpdateKaName(chapterObject.getString("comptkanaam"));
                        item.setUpdateKaLink(chapterObject.getString("referencelink"));
                        item.setUpdateKaDesc(chapterObject.getString("details"));
                        item.setSerialNumber(String.valueOf(i+1) + ".");
                        activity.updatesList.add(item);
                    }

                }else{
                    final String msg = jsonResponse.getString("message");
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(activity, msg, Toast.LENGTH_SHORT).show();
                        }
                    });
                }
            } catch (Exception e) { e.printStackTrace();}
            return null;
        }

        @Override
        protected void onPostExecute(JSONObject jsonObject) {
            final CompetitionUpdatesActivity activity = activityWeakReference.get();
            if(activity.pDialog.isShowing()) {
                activity.pDialog.dismiss();
            }
            activity.adapter = new CompetitionUpdatesAdapter(activity, activity.updatesList);
            activity.recyclerView.setAdapter(activity.adapter);
        }
    }
}

File: app/src/main/java/in/completecourse/questionbank/CompetitionUpdatesActivity.kt
package `in`.completecourse.questionbank

import `in`.completecourse.questionbank.adapter.CompetitionUpdatesAdapter
import `in`.completecourse.questionbank.app.AppConfig
import `in`.completecourse.questionbank.model.UpdateItem
import android.app.ProgressDialog
import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.activity_dialog.*
import org.json.JSONException
import org.json.JSONObject
import java.io.*
import java.lang.ref.WeakReference
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.*

class CompetitionUpdatesActivity : AppCompatActivity() {
    private var pDialog: ProgressDialog? = null
    private var updatesList: ArrayList<UpdateItem>? = null
    private var adapter: CompetitionUpdatesAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_dialog)

        recyclerView.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        updatesList = ArrayList<UpdateItem>()
        val classStringFinal = intent.getStringExtra("class")
        val dataObj = arrayOfNulls<String>(1)
        dataObj[0] = classStringFinal
        val getUpdates = GetUpdates(this@CompetitionUpdatesActivity)
        getUpdates.execute(*dataObj)

    }

    private class GetUpdates internal constructor(context: CompetitionUpdatesActivity) : AsyncTask<String?, String?, String?>() {
        private val activityWeakReference: WeakReference<CompetitionUpdatesActivity> = WeakReference(context)
        var item: UpdateItem? = null

        override fun onPreExecute() {
            val activity = activityWeakReference.get()
            activity!!.pDialog = ProgressDialog(activity)
            activity.pDialog!!.setMessage("Please wait...")
            activity.pDialog!!.setCancelable(false)
            activity.pDialog!!.show()
        }

        override fun doInBackground(vararg params: String?): String? {
            val activity = activityWeakReference.get()
            val urlString: String = AppConfig.URL_COMPETITION_UPDATES
            val studentclass = params[0]
            val url: URL
            val stream: InputStream
            var urlConnection: HttpURLConnection? = null
            try {
                url = URL(urlString)
                urlConnection = url.openConnection() as HttpURLConnection
                urlConnection.requestMethod = "POST"
                urlConnection.doOutput = true
                val data = (URLEncoder.encode("classid", "UTF-8") + "=" + URLEncoder.encode(studentclass, "UTF-8"))
                urlConnection.connect()
                val wr = OutputStreamWriter(urlConnection.outputStream)
                wr.write(data)
                wr.flush()
                stream = urlConnection.inputStream
                val reader = BufferedReader(InputStreamReader(stream, StandardCharsets.UTF_8), 8)
                val resFromServer = reader.readLine()
                val status: String
                val jsonResponse: JSONObject
                activity!!.updatesList = ArrayList<UpdateItem>()
                try {
                    jsonResponse = JSONObject(resFromServer)
                    //Log.e("chatpterItem", String.valueOf(jsonResponse));
                    status = jsonResponse.getString("status")
                    if (status == "true") {
                        val jsonObject = JSONObject(resFromServer)
                        val jsonArray = jsonObject.getJSONArray("data")
                        for (i in 0 until jsonArray.length()) {
                            item = UpdateItem()
                            val chapterObject = jsonArray.getJSONObject(i)
                            item!!.updateKaName = chapterObject.getString("comptkanaam")
                            item!!.updateKaLink = chapterObject.getString("referencelink")
                            item!!.updateKaDesc = chapterObject.getString("details")
                            item!!.serialNumber = (i + 1).toString() + "."
                            activity.updatesList!!.add(item!!)
                        }
                    } else {
                        val msg = jsonResponse.getString("message")
                        activity.runOnUiThread { Toast.makeText(activity, msg, Toast.LENGTH_SHORT).show() }
                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
                return null //reader.readLine();
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                urlConnection?.disconnect()
            }
            try {
                Thread.sleep(2000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
                Log.i("Result", "SLEEP ERROR")
            }
            return null
        }

        override fun onPostExecute(jsonObject: String?) {
            val activity = activityWeakReference.get()
            if (activity!!.pDialog!!.isShowing) {
                activity.pDialog!!.dismiss()
            }
            activity.adapter = CompetitionUpdatesAdapter(activity, activity.updatesList)
            activity.recyclerView!!.adapter = activity.adapter
        }

    }
}